### ISSUE TITLE
Unit tests for pagination

### ISSUE THREAD

**Issue by redshiftzero** [lucyparsons/OpenOversight#108]
Let's write some unit tests in for the pagination stuff cos I'm afraid to break it - we should do #102 first so that we have enough data to paginate (using the same number of officers per page as production).

**Comment by scuerda** [comment]
I spent a few moments this morning trying to get py.test set up for this and #102 but ran in to a number issues related to how flask configuration and test running and import paths. Not sure if this has been raised, but I've run in to this before w/ Flask apps. I am going to spend some more time trying to resolve this, though it might require a bit of reworking of how / when config is loaded.

**Comment by freddymartinez9** [comment]
@scuerda where you able to resolve your issue with testing ?

**Comment by scuerda** [comment]
@freddymartinez9 Sorry, I had to put this down for a bit...but no, I haven't made much headway.

pytest is a bit more finicky that unittest about paths and fixtures and without redoing a bunch of the application structure I'm having trouble getting pytest to be able to successfully start up an application instance. That said, I'll dive back in to the UnitTest test cases and see if I can get this done.

**Comment by scuerda** [comment]
@freddymartinez9 I've made a bit more headway, but I'm def in yak shaving territory right now.

The short of it is as follows. In order to setup pytest fixtures that allow for well scoped db transaction tests (including things like generating test data, etc) I had to do some serious application restructuring. The application is now using an app factory and a more structure config setup. 

I won't submit a PR yet, but [here](https://github.com/scuerda/OpenOversight/tree/generate-test-data) is the branch I'm working with.

I've def broken a lot of things at this point. Which maybe isn't a bad thing in terms of actually getting proper test coverage. But I'm just not sure how far down this path I should go. I strongly suspect that my refactoring will have completely broken the production deployment and without knowing how all that comes together I won't be of much use in terms of helping to fix that.

I'm happy to take some guidance on this if you or @redshiftzero have thoughts.

**Comment by scuerda** [comment]
I have this up to 91% coverage. What is currently outstanding is testing of the routes that accept post requests. As part of that next step I'll be adding in

- Arbitrary size (up to a limit) test set generation
- Explicit testing of the pagination

Currently I'm using in-memory sqlite for all of the tests. At some point I'll move it over to using postgresql to ensure reproducibility on the production stack.

**Comment by redshiftzero** [comment]
Just look at look at your branch - looks great! Major refactoring is fine - there are lots of nice improvements in there. We can figure out how to rework the deployment story using our staging server (set up in #104), but from looking at your branch the changes you are making shouldn't change the deployment story too much as we don't really have a smooth deploy process yet (pending in #38). We will have to rework parts of the development VM setup to use the fixtures for e.g. generating test data. Thanks for your work on this @scuerda!

**Comment by freddymartinez9** [comment]
Agreed with everything @redshiftzero. I looked at the diff just now and it looks great.

**Comment by scuerda** [comment]
@redshiftzero @freddymartinez9 ok, sounds good. I was looking at the [autogen-test-data](https://github.com/lucyparsons/OpenOversight/tree/autogen-test-data) branch and I think it makes sense for me to patch that in to the testing rig. In terms of the development vm, I don't think that there is much need rework how test data is loaded, except that it might make sense refactor the autogen functions into something that is referenced by both the test fixtures and the development test data. If we did that we could stick `NUM_OFFICERS` and the seed function in the `.env`. If that sounds good I'll take that approach when I patch in the autogen code.

**Comment by freddymartinez9** [comment]
We can close this yes @redshiftzero ?

**Comment by redshiftzero** [comment]
Nope we still need more tests for pagination
